Bluetooth: audio: pacs: Hotfix invalid PACS length

MariuszSkamra opened the pull request:
This fixes adding 2 redundant bytes... to the PAC records. As a result we end up with broken PAC record.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Hotfix invalid PACS length

This fixes adding 2 redundant bytes... to the PAC records.
As a result we end up with broken PAC record.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -2519,9 +2519,9 @@ static uint8_t unicast_client_read_func(struct bt_conn *conn, uint8_t err,
 
-		if (buf.len < sizeof(*pac)) {
+		if (buf.len < BT_PAC_RECORD_MIN_SIZE) {
 			BT_ERR("Malformed PAC: remaining len %u expected %zu",
-			       buf.len, sizeof(*pac));
+			       buf.len, BT_PAC_RECORD_MIN_SIZE);
```
I am not sure I understand this change. 

`sizeof(*pac)` was 7 before, right? 5 for the `codec`, 1 the `cc` and 1 for the `meta`. 

The new `BT_PAC_RECORD_MIN_SIZE` and `struct bt_pac` is now 
```c
sizeof(struct bt_pac) + 2 * sizeof(struct bt_pac_ltv_data)
```
which is 
```c
5 + 2 * 1
```

Which is the same value. 

The difference now is that `struct bt_pac` no longer contains a PAC record. 

Can you elaborate on this change?

MariuszSkamra replied:
It's true what you wrote, the difference is significant when you prepare net_buf memory in ASCS like here:
```
	struct bt_pac_ltv_data *cc, *meta;
	struct bt_pac *pac;

	pac = net_buf_simple_add(buf, sizeof(*pac));
	pac->codec.id = codec->id;
	pac->codec.cid = sys_cpu_to_le16(codec->cid);
	pac->codec.vid = sys_cpu_to_le16(codec->vid);

	cc = net_buf_simple_add(buf, sizeof(*cc));

	len = pac_data_add(buf, codec->data_count, codec->data);
	cc->len = len;
...
	meta = net_buf_simple_add(buf, sizeof(*meta));
```

As you can see it adds 7 bytes for PAC and then CC and meta as a result we get:

```
5 + 2 * 1 + 1 + 1
```

instead.

The problem is once we add `pac` so that `5 + 2 * 1` we can refer to pac->cc but pac->meta->data pointer will be incalid

Thalley replied:
Ah, yes, that makes sense. 

I guess we'll need to agree on what `bt_pac` semantically should include. The current definition
```c
struct bt_pac {
	struct bt_pac_codec codec;	/* Codec ID */
	struct bt_pac_ltv_data cc;	/* Codec Specific Capabilities */
	struct bt_pac_ltv_data meta;	/* Metadata */
} __packed;
```
doesn't make that much sense, as we actually never do `pac->cc` or `pac->meta`. 

Besides that, we shouldn't put a value like `bt_pac_ltv_data` that contains a `[0]` array in the middle of a struct. Empty arrays should always be at the end of a struct: https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html, which it both is, and isn't, for this particular struct. 

Given the uses of `bt_pac`, perhaps would should just remove that struct entirely and just use `bt_pac_codec` instead, which it, from the change in this PR, simply is?

MariuszSkamra replied:
Yeah, I'll remove it

MariuszSkamra committed:
Bluetooth: audio: unicast_client: Remove redundant code

The unicast_client_ep_set_codec function that is called below
sets the codec id, cid, vid.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Remove unused structure member

The 'pacs' member was never referenced.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
